Home Work 6

rickdemonn opened the pull request:
*Add Configuration yml
*Hello Hibernate
*Refactoring tests
*Refactoring Controller
*Refactoring Service
*Refactoring Repository to Spring Data

*in yml build list of specialization
*include dependency h2database
*test rewrite for Hibernate
*Controller refactoring for Hibernate and add Config and list of specializations
*Service refactoring for Hibernate
*Filters now in DB

Commit:
Home Work 6

*Add Configuration yml
*Hello Hibernate
*Refactoring tests
*Refactoring Controller
*Refactoring Service
*Refactoring Repository to Spring Data

*in yml build list of specialization
*include dependency h2database
*test rewrite for Hibernate
*Controller refactoring for Hibernate and add Config and list of specializations
*Service refactoring for Hibernate
*Filters now in DB

mfarsikov commented on src/main/java/hillel/spring/doctor/DoctorRestAPIController.java:
```diff
@@ -0,0 +1,99 @@
+
+    @PostMapping("/doctors")
+    public ResponseEntity<?> createDoctor(@RequestBody DoctorInputDto docDto){
+        final val doctor = doctorDtoConverter.toModel(docDto);
```
val is final by default

mfarsikov commented on src/main/java/hillel/spring/doctor/DoctorRestAPIController.java:
```diff
@@ -0,0 +1,99 @@
+            } catch (Exception e) {
+                throw new RuntimeException("Oops, error");
+            }
+        } else throw new WrongSpecializationsException();
```
Not important for this task, but nice to have: this exception will be logged, and it would be nice to have some explanation in logs: what actual value was, what allowed values are.

mfarsikov commented on src/main/java/hillel/spring/doctor/DoctorRestAPIController.java:
```diff
@@ -0,0 +1,99 @@
+                final val newDoc = doctorRestAPIService.createDoctor(doctor);
+                return ResponseEntity.created(uriComponentsBuilder.build(newDoc.getId())).build();
+            } catch (Exception e) {
+                throw new RuntimeException("Oops, error");
```
It is a really bad idea to swallow original exception: it contains a message and a stack trace. They can help find out what gone wrong.
Another bad idea to use this "Oops, error" it says nothing about the root cause, and it is better to avoid such messages.
So you should either wrap original exception or use lombok's `@SneakyThrows`  annotation

mfarsikov commented on src/main/java/hillel/spring/doctor/DoctorRestAPIController.java:
```diff
@@ -0,0 +1,99 @@
+
+@RestController
+public class DoctorRestAPIController {
+    private final DoctorRestAPIService doctorRestAPIService;
```
service should have nothing in common with REST API, so please rename service class and field

mfarsikov commented on src/main/java/hillel/spring/doctor/DoctorRestAPIRepo.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Query("select doctor from Doctor as doctor" +
+           " where doctor.specialization in :specializations")
+    List<Doctor> findBySpecializations(@Param("specializations") List<String> specializations);
```
try not using custom queries as much as you can: https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#jpa.query-methods.query-creation

mfarsikov commented on src/main/java/hillel/spring/doctor/DoctorRestAPIController.java:
```diff
@@ -0,0 +1,99 @@
+@RestController
+public class DoctorRestAPIController {
+    private final DoctorRestAPIService doctorRestAPIService;
+    private final DoctorDtoConverterImpl doctorDtoConverter;
```
there should be an interface instead

mfarsikov commented on src/main/java/hillel/spring/doctor/dto/DoctorDtoConverter.java:
```diff
@@ -0,0 +1,11 @@
+
+import hillel.spring.doctor.Doctor;
+
+//@Mapper
```
does it work?

mfarsikov commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,6 @@
+doctor.specializations.specializations:
+  - veterinarian
+  - surgeon
+  - geneticist
```
👍

mfarsikov commented on src/test/java/hillel/spring/doctor/DoctorRestAPIControllerTest.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    @Test
+    public void shouldCreateDoctor() throws Exception {
```
please add test for: cannot create a doctor with invalid specialization

mfarsikov commented on src/main/java/hillel/spring/doctor/dto/DoctorDtoConverterImpl.java:
```diff
@@ -0,0 +1,38 @@
+
+
+@Component
+public class DoctorDtoConverterImpl implements DoctorDtoConverter {
```
why this class is not generated?

Commit:
Home Work 6 fix

*rewrite MapStruct class
*rewrite Repo without @Query
*Rename DoctorController DoctorRepo DoctorService

Commit:
add dependency for MapStruct
